Add dnsforwardzone module

chr15p opened the pull request:
This PR adds an ipadnsforwardzone module to allow the creation of forwarders from the IPA dns server to external DNS servers. It follws the dnsforwardzone_* api calls as closely as possible.

t-woerner commented on plugins/modules/ipadnsforwardzone.py:
```diff
@@ -105,65 +106,80 @@ def find_dnsforwardzone(module, name):
+                        choices=["member", "dnsforwardzone"]),
             # state
-            state = dict(type='str', default='present', choices=['present', 'absent']),
+            state=dict(type='str', aliases=['ensure'], default='present',
```
Why is ensure an alias for state?

t-woerner commented on plugins/modules/ipadnsforwardzone.py:
```diff
@@ -180,35 +196,40 @@ def main():
-                    pass
+                    args = gen_args(forwarders, forwardpolicy,
+                                    skip_overlap_check)
+                    result = api_command(ansible_module, "dnsforwardzone_add",
```
result is assigned but never used.

t-woerner commented on plugins/modules/ipadnsforwardzone.py:
```diff
@@ -180,35 +196,40 @@ def main():
 
         except Exception as e:
-           # ansible_module.fail_json(msg="%s+ %s+ %s+ %s+ %s" % (command, name, res_find, args, compare_args_ipa(ansible_module, args, res_find), str(e)))
+            # ansible_module.fail_json(msg="%s+ %s+ %s+ %s" % (name, res_find, args,  str(e)))
```
This line is too long > 79, please split or remove if not needed.

t-woerner commented on plugins/modules/ipadnsforwardzone.py:
```diff
@@ -0,0 +1,246 @@
+short_description: Manage FreeIPA DNS Forwarder Zones
+description:
+- Add and delete an IPA DNS Forwarder Zones using IPA API
+options:
```
options is missing the ipaadmin_ parameters, please add them in the same way as other modules are doing this.

t-woerner commented on plugins/modules/ipadnsforwardzone.py:
```diff
@@ -0,0 +1,246 @@
+                                             "dnsforwardzone_mod",
+                                             name, args)
+                elif state == "absent":
+                    # exists but shouldn't
```
Why shouldn't the entry exist? With state: absent an existing entry should get removed.

t-woerner commented on plugins/modules/ipadnsforwardzone.py:
```diff
@@ -0,0 +1,246 @@
+        # Make sure forwardzone exists
+        res_find = find_dnsforwardzone(ansible_module, name)
+        if action == "member":
+            forwarders = list(set(res_find["idnsforwarders"] + forwarders))
```
With this line is not possible to make sure that entries are not part of idnsforwarders (`action: member` with `state: absent`). Also res_find is used without checking that it is not None.

t-woerner replied:
Please also add a test case to ensure the absence of an entry with also testing that changed is false for the repeated task.

t-woerner commented on tests/dnsforwardzone/test_dnsforwardzone.yml:
```diff
@@ -0,0 +1,136 @@
+    ipadnsforwardzone:
+      ipaadmin_password: password01
+      name: example.com
+      ensure: absent
```
This should be `state`

t-woerner commented on tests/dnsforwardzone/test_dnsforwardzone.yml:
```diff
@@ -0,0 +1,136 @@
+        - 8.8.8.8
+      forwardpolicy: first
+      skip_overlap_check: true
+      ensure: present
```
...

t-woerner commented on tests/dnsforwardzone/test_dnsforwardzone.yml:
```diff
@@ -0,0 +1,136 @@
+        - 8.8.8.8
+      forwardpolicy: first
+      skip_overlap_check: false 
+      ensure: present
```
...

t-woerner commented on tests/dnsforwardzone/test_dnsforwardzone.yml:
```diff
@@ -0,0 +1,136 @@
+        - 4.4.4.4
+      forwardpolicy: only
+      skip_overlap_check: false 
+      ensure: present
```
...

t-woerner replied:
You are using `state` and `ensure` here.. that results in a warning: `[WARNING]: Both option state and its alias ensure are set.`

t-woerner commented on tests/dnsforwardzone/test_dnsforwardzone.yml:
```diff
@@ -0,0 +1,136 @@
+    ipadnsforwardzone:
+      ipaadmin_password: password01
+      name: example.com
+      ensure: absent
```
...

t-woerner commented on tests/dnsforwardzone/test_dnsforwardzone.yml:
```diff
@@ -0,0 +1,136 @@
+    ipadnsforwardzone:
+      ipaadmin_password: password01
+      name: example.com
+      ensure: absent
```
...

t-woerner commented on README-dnsforwardzone.md:
```diff
@@ -0,0 +1,106 @@
+    ipadnsforwardzone:
+      ipaadmin_password: password01
+      name: example.com
+      ensure: absent
```
`ensure` parameter is also used here.

t-woerner commented on README.md:
```diff
@@ -403,6 +403,7 @@ Roles
 Modules in plugin/modules
```
Please also add a line to `Features` in the first section of the README.

t-woerner commented:
There are also white space changes reported by git: 
```
Applying: add a module to manage dns forwarder zones in ipa
.git/rebase-apply/patch:44: trailing whitespace.
author: chris procter 
.git/rebase-apply/patch:58: trailing whitespace.
  forwarders: 
.git/rebase-apply/patch:60: trailing whitespace.
    - The DNS servers to forward to 
.git/rebase-apply/patch:63: trailing whitespace.
    description: Per-zone conditional forwarding policy 
.git/rebase-apply/patch:67: trailing whitespace.
  skip_overlap_check: 
warning: squelched 3 whitespace errors
warning: 8 lines add whitespace errors.
Applying: cleanup following flake8, pytest, and comments from developer
Applying: add tests for dnsforwardzone
.git/rebase-apply/patch:75: trailing whitespace.
  - name: skip_overlap_check can only be set on creation so do not change when its changed 
.git/rebase-apply/patch:83: trailing whitespace.
      skip_overlap_check: false 
.git/rebase-apply/patch:97: trailing whitespace.
      skip_overlap_check: false 
.git/rebase-apply/patch:147: new blank line at EOF.
+
warning: 4 lines add whitespace errors.
Applying: add documentation for dnsforwardzone
```

chr15p commented:
After talking to t-woerner IRL and reading the comments I have tidied up, added the enabled/disabled states, and rebased my changes into a single commit. Hopefully this addresses all the issues (and probably adds lots of new ones :)

t-woerner commented on README-dnsforwardzone.md:
```diff
@@ -0,0 +1,112 @@
+`ipaadmin_password` | The admin password is a string and is required if there is no admin ticket available on the node | no
+`name` \| `cn` | Zone name (FQDN). | yes if `state` == `present`
+`forwarders` \| `idnsforwarders` |  Per-zone conditional forwarding policy. Possible values are `only`, `first`, `none`) | no
+`forwardpolicy` \| `idnsforwardpolicy` |
```
This is missing a description and also the required field.

t-woerner commented on README-dnsforwardzone.md:
````diff
@@ -0,0 +1,112 @@
+```
+
+
+Example playbook to add a forwardzone to ipa DNS:
````
This should be more like "ensure presence of ....". It would also be nice to cut the playbook into smaller ones. 1) "forward DNS requests for example.com to 8.8.8.8", 2) "ensure forward zone is disabled", 3) ...

t-woerner commented on README-dnsforwardzone.md:
```diff
@@ -0,0 +1,112 @@
+      name: example.com
+      state: disabled
+
+  - name: add multiple upstream DNS servers for example.com
```
This should also be more like "ensure presence of .. in .."

t-woerner commented on README-dnsforwardzone.md:
```diff
@@ -0,0 +1,112 @@
+      forwardpolicy: first
+      skip_overlap_check: true
+
+  - name: disable the forward zone
```
This should be more like "ensure forward zone is disabled"

t-woerner commented on README-dnsforwardzone.md:
```diff
@@ -0,0 +1,112 @@
+        - 8.8.8.8
+        - 4.4.4.4
+
+  - name: add another forwarder to any existing ones for example.com
```
...

scaronni commented:
I would be very happy seeing a module for DNS forward zones being merged. In the meanwhile, using the module I get this error for the following domains:

```
failed: [ipa01.example.com] (item=xn--dnscheckswiss-4nbs3b5b4g9hxe4d.com) => {"ansible_loop_var": "item", "changed": false, "item": "xn--dnscheckswiss-4nbs3b5b4g9hxe4d.com", "msg": "('IDNA does not round-trip', 'xn--dnscheckswiss-4nbs3b5b4g9hxe4d', 'xn--ssdnscheckswiss-4tb5b7b9gwi2e8d')"}
failed: [ipa01.example.com] (item=xn--htmlcheckswiss-uqbt5b7b9gwi2e8d.com) => {"ansible_loop_var": "item", "changed": false, "item": "xn--htmlcheckswiss-uqbt5b7b9gwi2e8d.com", "msg": "('IDNA does not round-trip', 'xn--htmlcheckswiss-uqbt5b7b9gwi2e8d', 'xn--sshtmlcheckswiss-vwb7b9b1hwi3e9d')"}
```

We do CAA/DNSSEC validation on zones, so we have all sort of zones including Unicode / Internationalized domain names for testing. See https://en.wikipedia.org/wiki/Internationalized_domain_name for details.

chr15p commented:
> failed: [ipa01.example.com] (item=xn--dnscheckswiss-4nbs3b5b4g9hxe4d.com) => {"ansible_loop_var": "item", "changed": false, "item": "xn--dnscheckswiss-4nbs3b5b4g9hxe4d.com", "msg": "('IDNA does not round-trip', 'xn--dnscheckswiss-4nbs3b5b4g9hxe4d', 'xn--ssdnscheckswiss-4tb5b7b9gwi2e8d')"}
> failed: [ipa01.example.com] (item=xn--htmlcheckswiss-uqbt5b7b9gwi2e8d.com) => {"ansible_loop_var": "item", "changed": false, "item": "xn--htmlcheckswiss-uqbt5b7b9gwi2e8d.com", "msg": "('IDNA does not round-trip', 'xn--htmlcheckswiss-uqbt5b7b9gwi2e8d', 'xn--sshtmlcheckswiss-vwb7b9b1hwi3e9d')"}

can you post the ansible task you are running, I'll have a look and see if I can figure out where the problem is

chr15p commented:
@t-woerner can I get a code review please

t-woerner commented:
There is an additional parameter that should be added: `permission` for "permission for per forward zone access delegation". See `dnsforwardzone-add-permission` and `dnsforwardzone-remove-permission` commands.

t-woerner commented on tests/dnsforwardzone/test_dnsforwardzone.yml:
```diff
@@ -0,0 +1,214 @@
+---
+- name: Tests
```
Please change this to "Test dnsforwardzone".

rjeffman commented on tests/dnsforwardzone/test_dnsforwardzone.yml:
```diff
@@ -0,0 +1,214 @@
+    register: result
+    failed_when: not result.changed
+
+  - name: iensure it stays disabled
```
'iensure' should be 'ensure', right? (Small typo).

chr15p replied:
the curse of using vim :)

rjeffman replied:
It could be worse if using emacs. :)

rjeffman commented:
@chr15p you are still missing one [documentation change](https://github.com/freeipa/ansible-freeipa/pull/163/commits/17bc99107a04dec44a60fbfd3c9d490c723028fe#r357205163), otherwise, the patch LGTM.

t-woerner commented:
LGTM
